server/src/main/java/io/crate/replication/logical/ShardTranslogSequencer.java

```diff
@@ -0,0 +1,109 @@
+import org.elasticsearch.index.shard.ShardId;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A TranslogSequencer allows multiple producers of [Translog.Operation]s to write them in sequence number order to an
+ * index. Producer can call the [send] method to add a batch of operations to the queue. Operations can be
+ * sent out of order i.e. the operation with sequence number 2 can be sent before the operation with sequence number 1.
+ * In this case the Sequencer will internally buffer the operations that cannot be delivered until the missing in-order
+ * operations arrive.
+ *
+ * Derived from org.opensearch.replication.task.shard.TranslogSequencer
+ */
+public class ShardTranslogSequencer {
+
+    private static final Logger LOGGER = Loggers.getLogger(ShardTranslogSequencer.class);
+
+    private final ShardId shardId;
+    private final Client client;
+    private final long initialSeqNo;
+    private final ConcurrentHashMap<Long, ShardChangesAction.Response> unAppliedChanges = new ConcurrentHashMap<>();
+
+    public ShardTranslogSequencer(ShardId shardId,
+                                  Client client,
+                                  long initialSeqNo) {
+        this.shardId = shardId;
+        this.client = client;
+        this.initialSeqNo = initialSeqNo;
+    }
+
+    public void send(ShardChangesAction.Response response) throws Exception {
+        unAppliedChanges.put(response.fromSeqNo(), response);
+        process();
+    }
+
+    private synchronized void process() throws Exception {
+        AtomicReference<Exception> lastError = new AtomicReference<>();
+        var highWatermark = initialSeqNo;
+        while (unAppliedChanges.containsKey(highWatermark + 1) && lastError.get() == null) {
+            var next = unAppliedChanges.remove(highWatermark + 1);
+            var changes = next.changes();
+            if (changes.size() > 0) {
+                var replayRequest = new ReplayChangesAction.Request(
+                    shardId,
+                    next.changes(),
+                    next.maxSeqNoOfUpdatesOrDeletes()
+                );
+                client.execute(
+                    ReplayChangesAction.INSTANCE,
+                    replayRequest,
+                    new ActionListener<>() {
+                        @Override
+                        public void onResponse(ReplayChangesAction.Response replayResponse) {
+                            var shardInfo = replayResponse.getShardInfo();
+                            if (shardInfo.getFailed() > 0) {
+                                for (ReplicationResponse.ShardInfo.Failure failure : shardInfo.getFailures()) {
+                                    LOGGER.error("Failed replaying changes. Failure:{}", failure);
+                                }
+                                lastError.set(new RuntimeException("Some changed failed while replaying"));
```

Comment:
`lastError` is set asynchronously here in the callback, but accessed after the loop and within the while condition. Is this really working as intended?